geode-core/src/main/java/org/apache/geode/internal/datasource/GemFireTransactionDataSource.java

```diff
@@ -219,7 +221,13 @@ void registerTranxConnection(XAConnection xaConn) throws Exception {
    * @return ???
    */
   protected Connection getSQLConnection(PooledConnection poolC) throws SQLException {
-    Connection conn = poolC.getConnection();
+    Connection conn;
+    try {
+      conn = poolC.getConnection();
+    } catch (SQLException e) {
+      provider.returnAndExpireConnection(poolC);
+      throw new SQLException(e.toString());
```

Comment:
I would prefer to use the constructor that takes another exception (that way we can retain the full stack trace).